File: smb:/192.168.99.130/staging/Release/System/API/Permission.js
require("./PermissionDef");
var PermissionTable = {};
function SetPermission(id, buffer) {
    PermissionTable[id] = Encode(Decode(buffer));
}
exports.SetPermission = SetPermission;
function GetPermission(id) {
    return PermissionTable[id];
}
exports.GetPermission = GetPermission;
var _permission_buffer_length = Math.ceil((Object.keys(Permission).length / 2) / 32);
var _perm_length = (Object.keys(Permission).length / 2);
var _GROUP_SYSTEM = _.range(_perm_length);
function Resolve(raw_p) {
    if (!raw_p)
        return undefined;
    for (var i = 0; i < raw_p.length; i++) {
        switch (raw_p[i]) {
            case 0 /* System */:
                return (_GROUP_SYSTEM);
                break;
            default:
                break;
        }
    }
    return raw_p;
}
function Encode(permissions) {
    if (!permissions)
        return undefined;
    permissions = Resolve(permissions);
    var buf = new Array(_permission_buffer_length);
    for (var i = 0; i < buf.length; i++) {
        buf[i] = 0;
    }
    for (var i = 0; i < permissions.length; i++) {
        var bitIndex = permissions[i];
        buf[Math.floor(bitIndex / 32)] |= 1 << (bitIndex % 32);
    }
    return buf;
}
exports.Encode = Encode;
function Decode(permissions) {
    var p = [];
    if (!permissions)
        return p;
    for (var i = 0; i < permissions.length; i++) {
        for (var j = 0; j < 32; j++) {
            if ((permissions[i] & (1 << j)) != 0) {
                p.push(Permission[Permission[i * 32 + j]]);
            }
        }
    }
    return p;
}
exports.Decode = Decode;
function DecodeToString(permissions) {
    var p = [];
    if (!permissions)
        return p;
    for (var i = 0; i < permissions.length; i++) {
        for (var j = 0; j < 32; j++) {
            if ((permissions[i] & (1 << j)) != 0) {
                p.push(Permission[i * 32 + j]);
            }
        }
    }
    return p;
}
exports.DecodeToString = DecodeToString;
function Check(owned, required) {
    if (!required) {
        return true;
    }
    for (var i = 0; i < required.length; i++) {
        if (required[i] == 0)
            continue;
        if (!owned || i >= owned.length)
            return false;
        if ((owned[i] & required[i]) != required[i])
            return false;
    }
    return true;
}
exports.Check = Check;

File: System/Scripts/Tools/API/permission.lua
local bit = require 'bit'
local PermissionTable = {}

PermissionOps = {}

Permission = {
    All = 0,
    Proxy = 1,
    Event = 2,
    IO = 3,
    DeviceAccess = 4,
    UserAccess = 5,
    Sensor = 6,
    AnyApp = 7,
    AppCrossTalk = 8,
    AppPreLaunch = 9,
    PortExposure = 10,
    Network = 11,
    Configuration = 12,
    Launcher = 13,
    Driver = 14,
}

local PermissionCount = 0
for _ in pairs(Permission) do
    PermissionCount = PermissionCount + 1
end

local PermArrLength = math.ceil(PermissionCount / 32)

function PermissionOps.Encode(Perms)
    if (not Perms) then return nil end

    local permArr = { n = PermArrLength }

    for i = 1, PermArrLength do
        permArr[i] = 0
    end

    for i = 1, #Perms do
        local bitIndex = Perms[i] -- 0 ~ N
        permArr[math.floor(bitIndex / 32) + 1] =
        bit.bor(permArr[math.floor(bitIndex / 32) + 1],
            bit.lshift(1, bitIndex % 32))
        --buf[Math.floor(bitIndex / 32)] |= 1 << (bitIndex % 32); //flip bit
    end

    return permArr
end

function PermissionOps.Set(id, Perm)
    PermissionTable[id] = PermissionOps.Encode(Perm)
end

function PermissionOps.Get(id)
    return PermissionTable[id]
end

function PermissionOps.Check(owned, required)
    if (not required) then return true end
    if (owned and owned[1] and bit.band(owned[1], 1) == 1) then return true end --system
    for i = 1, #required do
        if (required[i] ~= 0 and (owned == nil or owned[i] == nil or owned[i] == 0)) then
            return false
        elseif ( bit.band(owned[i], required[i]) ~= required[i] ) then
            return false
        end
    end
    return true
end

--local s = PermissionOps.Encode({Permission.All})
--local a = PermissionOps.Encode({Permission.Proxy, Permission.Sensor})
--local b = PermissionOps.Encode({Permission.Proxy, Permission.Sensor, Permission.AnyApp})
--
--print(PermissionOps.Check(a, b))    --f
--print(PermissionOps.Check(b, a))    --t
--print(PermissionOps.Check(nil, a))  --f
--print(PermissionOps.Check(a, nil))  --t
--print(PermissionOps.Check(s, a))    --t